### PR TITLE
feat: assign a dynamic port to both front and back ends on all ui project

### PR DESCRIPTION
**Pull request by y-lakhdar**
## Proposed changes

Prevent the application to crash if the specified ports are already in use. Instead, the next available ports will be held for both the front-end application and search token server.
The idea is to run a script (`script/port-allocator.js`) upstream to allocate available ports to the front and back. The script checks if a port has already been specified by the user.
In all cases, a validation ensures that the port is actually available.

For React and Vue, it’s pretty straightforward: Run the port-allocator logic and hold the available port in a .env file. However, it is getting a little bit trickier with Angular.

Angular uses `environment.ts` files to store environment variables. It becomes very difficult (and against the Angular best practices) to read values from an `.env` file. In this case, the port-allocator service will directly read and update the `environment.ts` file with the appropriate port to the search token server.

There is a very tiny chance of a race condition if another process starts using the same port number as you in between the time you get the port number and you actually start using it... but seriously...

## Testing

- [ ] E2E Tests: Add a test in which the application tries to run on a busy port
- [x] Manual Tests:
    Tested with the 3 frameworks. Whenever the web application (or search token server) tries to run on a busy port, they get allocated available ports.

-----
CDX-99

### CODE REVIEW

File: packages/cli-e2e/__tests__/vue.specs.ts [hunk 1]
```diff
@@ -52,22 +97,37 @@ describe('ui:create:vue', () => {
...
+  ) => {
     const serverTerminal = new Terminal(
       'npm',
       ['run', 'start'],
       {
         cwd: getProjectPath(projectName),
       },
       processManager,
-      'vue-server'
+      debugName
```
Comment (y-lakhdar) [inline packages/cli-e2e/__tests__/vue.specs.ts:112]: Otherwise we have multiple `startApplication` Terminals writing in the same stderr and stdout files

File: packages/cli-e2e/utils/browser.ts [hunk 2]
```diff
@@ -245,7 +245,7 @@ export async function isElementClickable(page: Page, selector: string) {
           isElementInViewport(elem) &&
           elem.disabled !== true &&
           isOverlappingElementMatch(
-            getOverlappingElements(elem) as any as HTMLElement[],
+            (getOverlappingElements(elem) as any) as HTMLElement[],
```
Comment (y-lakhdar) [inline packages/cli-e2e/utils/browser.ts:248]: prettier

File: packages/angular/src/search-token-server/files/scripts/port-allocator.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+const getPort = require('get-port');
+const {config} = require('dotenv');
+const pathToServerEnv = resolve('server', '.env');
+const environment = config({path: pathToServerEnv}).parsed;
+
+const preferedWebAppPort = 4200;
+const portRangeFallback = [4000, 4999];
+
+const ensureEnvironmentFile = () => {
+  if (!existsSync(pathToServerEnv)) {
```
Comment (y-lakhdar) [inline packages/angular/src/search-token-server/files/scripts/port-allocator.js:19]: Let's stop the process immediately if the `.env` file is missing instead of running an invalid project.
It also helps with the E2E tests.

**Review by louis-bompart — COMMENTED** [review]
first review, second ongoing :loading:

**Review by louis-bompart — COMMENTED** [review]
Looks good overall GG!

File: packages/angular/src/ng-add-setup-project/files/src/environments/environment.ts.template [hunk 4]
```diff
@@ -4,8 +4,25 @@
 
 export const environment = {
   production: false,
-  tokenEndpoint: '/token',
+  /**
+   * The endpoint to the server returning Coveo search tokens.
+   * Example: tokenEndpoint: 'https://acme.com/token',
+   * If kept to null, `defaultTokenEndpoint` will be used fetch the search token from the local server within this project
+   */
+  customtokenEndpoint: null,
```
[the remark this answers is not in the document]
Comment (olamothe) [reply packages/angular/src/ng-add-setup-project/files/src/environments/environment.ts.template:12]: Undefined would be better here.

**Review by olamothe — APPROVED** [review]
It's quite hard to review, not going to lie...

For next time, when you see those kind of things happening (a PR thats keeps getting bigger and bigger), please split it logically.

Either adding tests in a separate PR, or doing each framework (vue/angular/react) as separate, things like that.

More than couple hundred LOC changed, and it gets pretty hard to follow.


On a separate not, instead of writing to an .env file, would an alternative be to * always * make this dynamic ?

Like, for each project, we inject a "start a server" node script, which runs the correct command for each respective project, but that determine an available port before doing so ?

File: packages/angular/src/ng-add-setup-project/rules/start.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import {Rule, SchematicContext, Tree} from '@angular-devkit/schematics';
+import {CoveoSchema} from '../../schema';
+
+/**
+ * Create a npm script in the project root to start both the Angular app and the proxy server
+ */
+export function configureStartCommand(_options: CoveoSchema): Rule {
+  return (tree: Tree, _context: SchematicContext) => {
+    const packageJsonBuffer = tree.read(normalize('./package.json'));
+    if (packageJsonBuffer === null) {
```
[the remark this answers is not in the document]
Comment (y-lakhdar) [reply packages/angular/src/ng-add-setup-project/rules/start.ts:11]: Actually no. I'll remove it. there is no point in that check. if something breaks, it will happen in the try catch

[the remark this answers is not in the document]
Comment (y-lakhdar) [reply packages/angular/src/ng-add-setup-project/rules/start.ts]: Oh now I remember. It returns a TS error otherwise
image

**Comment by y-lakhdar** [comment]
> Like, for each project, we inject a "start a server" node script, which runs the correct "server" command for each respective project, but that determine an available port before doing so ?

That's what the port-allocator.js file is doing. The reason I did not put that logic inside the start-server script is because the Web app needs to know the token endpoint, hence the `.env` file.

**Comment by olamothe** [comment]
> > Like, for each project, we inject a "start a server" node script, which runs the correct "server" command for each respective project, but that determine an available port before doing so ?
> 
> That's what the port-allocator.js file is doing. The reason I did not put that logic inside the start-server script is because the Web app needs to know the token endpoint, hence the `.env` file.

👍  Got it !
